server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -556,11 +569,19 @@ void recoverFilesFromSourceAndSnapshot(ShardRecoveryPlan shardRecoveryPlan,
             sendFileInfoStep
         );
 
-        final List<StoreFileMetadata> sourceFiles = shardRecoveryPlan.getSourceFilesToRecover();
+        sendFileInfoStep.whenComplete(r -> recoverSnapshotFiles(shardRecoveryPlan, recoverSnapshotFilesStep), listener::onFailure);
+
+        recoverSnapshotFilesStep.whenComplete(filesFailedToRecoverFromSnapshot -> {
+            final List<StoreFileMetadata> filesToRecoverFromSource;
+            if (filesFailedToRecoverFromSnapshot.isEmpty()) {
+                filesToRecoverFromSource = shardRecoveryPlan.getSourceFilesToRecover();
+            } else {
+                filesToRecoverFromSource = concatLists(shardRecoveryPlan.getSourceFilesToRecover(), filesFailedToRecoverFromSnapshot);
```

Comment:
I would prefer to let the primary keep track of what has been downloaded successfully and act accordingly. About the case you mention, I think that we should introduce a new step there in case of failures and clear out everything that has been downloaded since we cannot reuse any of the downloaded files and let the primary to send its index files.

Reply:
OK, some of the other failure handling comments clarified this for me, the approach looks ok.